examples/demo-stf/src/batch_builder.rs

```diff
@@ -0,0 +1,332 @@
+        Self {
+            mempool: RefCell::new(VecDeque::new()),
+            mempool_max_txs_count,
+            batch_size_bytes,
+            runtime,
+            working_set: None,
+        }
+    }
+
+    #[cfg(feature = "native")]
+    pub fn reset_working_set(&mut self, working_set: WorkingSet<<C as Spec>::Storage>) {
+        self.working_set = Some(RefCell::new(working_set));
+    }
+}
+
+impl<R, C: Context> BatchBuilder for FiFoStrictBatchBuilder<R, C>
+where
+    R: DispatchCall<Context = C>,
+{
+    /// Transaction can only be declined only mempool is full
+    fn accept_tx(&self, tx: Vec<u8>) -> anyhow::Result<()> {
+        if self.mempool.borrow().len() > self.mempool_max_txs_count {
+            bail!("Mempool is full")
+        }
+        let mut mempool = self.mempool.borrow_mut();
+        mempool.push_back(tx);
+        Ok(())
+    }
+
+    /// Builds a new batch of valid transactions in order they were added to mempool
+    /// Only transactions, which are dispatched successfully are included in the batch
+    fn get_next_blob(&self) -> anyhow::Result<Vec<Vec<u8>>> {
+        let working_set = match self.working_set.as_ref() {
+            None => {
+                bail!("Cannot build batch before working set is initialized");
+            }
+            Some(working_set) => working_set,
+        };
+        let mut txs = Vec::new();
+        let mut dismissed: Vec<(Vec<u8>, anyhow::Error)> = Vec::new();
+        let mut current_batch_size = 0;
+        let mut mempool = self.mempool.borrow_mut();
+
+        while let Some(raw_tx) = mempool.pop_front() {
+            let tx_len = raw_tx.len();
+
+            // Deserialize
+            let mut data = Cursor::new(&raw_tx);
+            let tx = match Transaction::<C>::deserialize_reader(&mut data) {
+                Ok(tx) => tx,
+                Err(err) => {
+                    let err = anyhow::Error::new(err).context("Failed to deserialize transaction");
+                    dismissed.push((raw_tx, err));
+                    continue;
+                }
+            };
+
+            // Verify
+            if let Err(err) = tx.verify() {
+                dismissed.push((raw_tx, err));
+                continue;
+            }
+
+            // Decode
+            let msg = match R::decode_call(tx.runtime_msg()) {
+                Ok(msg) => msg,
+                Err(err) => {
+                    let err =
+                        anyhow::Error::new(err).context("Failed to decode message in transaction");
+                    dismissed.push((raw_tx, err));
+                    continue;
+                }
+            };
+
+            // Execute
+            {
+                let sender_address: C::Address = tx.pub_key().to_address();
+                let ctx = C::new(sender_address);
+                let mut working_set = working_set.borrow_mut();
+
+                //
+                match self.runtime.dispatch_call(msg, &mut working_set, &ctx) {
```

Comment:
In my opinion we should allow transactions that throws error when dispatched to appropriate module.

Reply:
This implementation is opinionated and not considered standard.

Main purpose of this contraint is to add runtime, to illustrate how BatchBuilder can be used against state